src/null-safety/understanding-null-safety/index.md

```diff
@@ -0,0 +1,1410 @@
+
+In practice, this means:
+
+*   If you want to indicate that you allow a value of any type, use `Object?`
+    instead of `Object`. In fact, it becomes pretty rare to use `Object` since
+    that type means "could be any possible value except this one weirdly
+    prohibted value `null`".
+
+*   On the rare times you need a bottom type, use `Never` instead of `Null`.
+    If you don't know if you need it, you probably don't.
+
+## Ensuring correctness
+
+We bifurcated the universe of types into nullable and non-nullable halves. In
+order to maintain soundness and our principle that you can never get a null
+reference error at runtime unless you ask for it, we need to guarantee that
+`null` never appears in any type on the non-nullable side.
+
+Getting rid of implicit downcasts and removing `Null` as a bottom type covers
+all of the main places that types flow through a program across assigments and
+from arguments into parameters on function calls. The main remaining places
+where `null` can sneak in are when a variable first comes into being and when
+you leave a function. So there are some additional compile errors:
+
+### Invalid returns
+
+If a function has a non-nullable return type, then every path through the
+function must reach a `return` statement that returns a value. Before null
+safety, Dart was pretty lax about missing returns. For example:
+
+```dart
+String missingReturn() {
+  // No return.
+}
+```
+
+If you analyze this, you get a gentle *hint* that *maybe* you forgot a return,
+but if not, no big deal. That's because if execution reaches the end of a
+function body then Dart implicitly returns `null`. Since every type is nullable,
+*technically* this function is safe, even though it's probably not what you
+want.
+
+With sound non-nullable types, this program is flat out wrong and unsafe. Under
+null safety, you get a compile error if a function with a non-nullable return
+type doesn't reliably return a value. By "reliably", I mean that the language
+analyzes all of the control flow paths through the function. As long as they all
+return something, it is satisfied. The analysis is pretty smart, so even this
+function is OK:
+
+```dart
+String alwaysReturns(int n) {
+  if (n == 0) {
+    return 'zero';
+  } else if (n < 0) {
+    throw ArgumentError('Negative values not allowed.');
+  } else {
+    if (n > 1000) {
+      return 'big';
+    } else {
+      return n.toString();
+    }
+  }
+}
+```
+
+We'll dive more deeply into the new flow analysis in the next section.
+
+### Uninitialized variables
+
+When you declare a variable, if you don't give it an explicit initializer, Dart
+default initializes the variable with `null`. That's convenient, but obviously
+totally unsafe if the variable's type is non-nullable. So we have to tighten
+things up for non-nullable variables:
+
+*   **Top level variable and static field declarations must have an
+    initializer.** Since these are lazy and can be accessed and assigned from
```

Comment:
[nit] If there's a place to link to from "lazy", I'd do it. Not everyone will know what you mean.

Reply:
Actually, laziness doesn't come into play anyway. Deleted.